subsys/net/lib/nrf_cloud/coap/src/nrf_cloud_coap.c

```diff
@@ -0,0 +1,489 @@
+		[NRF_CLOUD_FOTA_REJECTED]    = "REJECTED",
+		[NRF_CLOUD_FOTA_CANCELED]    = "CANCELLED",
+		[NRF_CLOUD_FOTA_DOWNLOADING] = "DOWNLOADING",
+	};
+#define JOB_STATUS_STRING_COUNT (sizeof(job_status_strings) / \
+				 sizeof(*job_status_strings))
+
+	int ret;
+	size_t buff_sz;
+	char *url = NULL;
+	char *payload = NULL;
+
+	/* Format API URL with device and job ID */
+	buff_sz = sizeof(API_UPDATE_FOTA_URL_TEMPLATE) + strlen(job_id);
+	url = nrf_cloud_malloc(buff_sz);
+	if (!url) {
+		ret = -ENOMEM;
+		goto clean_up;
+	}
+
+	ret = snprintk(url, buff_sz, API_UPDATE_FOTA_URL_TEMPLATE, job_id);
+	if ((ret < 0) || (ret >= buff_sz)) {
+		LOG_ERR("Could not format URL");
+		ret = -ETXTBSY;
+		goto clean_up;
+	}
+
+	/* Format payload */
+	if (details) {
+		buff_sz = sizeof(API_UPDATE_FOTA_DETAILS_TMPLT) +
+			  strlen(job_status_strings[status]) +
+			  strlen(details);
+	} else {
+		buff_sz = sizeof(API_UPDATE_FOTA_BODY_TEMPLATE) +
+			  strlen(job_status_strings[status]);
+	}
+
+	payload = nrf_cloud_malloc(buff_sz);
+	if (!payload) {
+		ret = -ENOMEM;
+		goto clean_up;
+	}
+
+	if (details) {
+		ret = snprintk(payload, buff_sz, API_UPDATE_FOTA_DETAILS_TMPLT,
+			       job_status_strings[status], details);
+	} else {
+		ret = snprintk(payload, buff_sz, API_UPDATE_FOTA_BODY_TEMPLATE,
+			       job_status_strings[status]);
+	}
+	if ((ret < 0) || (ret >= buff_sz)) {
+		LOG_ERR("Could not format payload");
+		ret = -ETXTBSY;
+		goto clean_up;
+	}
+
+	ret = nrf_cloud_coap_patch(url, NULL, payload, ret,
+				   COAP_CONTENT_FORMAT_APP_JSON, true, NULL, NULL);
+
+clean_up:
+	if (url) {
+		nrf_cloud_free(url);
+	}
+	if (payload) {
+		nrf_cloud_free(payload);
+	}
+
+	return ret;
+}
+
+struct get_shadow_data  {
+	char *buf;
+	size_t buf_len;
+} get_shadow_data;
+static int shadow_err;
+
+static void get_shadow_callback(int16_t result_code,
+		       size_t offset, const uint8_t *payload, size_t len,
+		       bool last_block, void *user)
+{
+	struct get_shadow_data *data = (struct get_shadow_data *)user;
+
+	LOG_DBG("result_code: %d.%02d, offset:0x%X, len:0x%X, last_block:%d",
+		result_code / 32u, result_code & 0x1f, offset, len, last_block);
+	if (result_code != COAP_RESPONSE_CODE_CONTENT) {
+		shadow_err = result_code;
+	} else {
+		int cpy_len = MIN(data->buf_len - 1, len);
+
+		shadow_err = 0;
+		memcpy(data->buf, payload, cpy_len);
+		data->buf[len] = '\0';
+	}
+}
+
+int nrf_cloud_coap_shadow_get(char *buf, size_t buf_len, bool delta)
+{
+	__ASSERT_NO_MSG(buf != NULL);
+	CHECK_AUTH();
+
+	get_shadow_data.buf = buf;
+	get_shadow_data.buf_len = buf_len;
+
+	return nrf_cloud_coap_get("state", delta ? NULL : "delta=false", NULL, 0,
+				    0, COAP_CONTENT_FORMAT_APP_JSON, true, get_shadow_callback,
+				    &get_shadow_data);
+}
+
+int nrf_cloud_coap_shadow_state_update(const char * const shadow_json)
+{
+	__ASSERT_NO_MSG(shadow_json != NULL);
+	CHECK_AUTH();
+
+	return nrf_cloud_coap_patch("state", NULL, (uint8_t *)shadow_json,
+				    strlen(shadow_json),
```

Comment:
any size limitations when sending coap data?

Reply:
Yes, `CONFIG_COAP_CLIENT_BLOCK_SIZE` which can range from 64 to 1024.

Reply:
Actually, it's fine.  The coap_client library in Zephyr automatically detects data larger than the maximum size of a single message, and breaks it up into multiple CoAP block transfers.